common/archiver/gcloud/historyArchiver.go

```diff
@@ -0,0 +1,404 @@
+			return errUploadNonRetriable
+		}
+
+		filename := constructHistoryFilenameMultipart(request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion, part)
+		if exist, _ := h.gcloudStorage.Exist(ctx, URI, filename); !exist {
+			if err := h.gcloudStorage.Upload(ctx, URI, filename, encodedHistoryPart); err != nil {
+				logger.Error(archiver.ArchiveTransientErrorMsg, tag.ArchivalArchiveFailReason(errWriteFile), tag.Error(err))
+				scope.IncCounter(metrics.HistoryArchiverArchiveTransientErrorCount)
+				return err
+			}
+
+			totalUploadSize = totalUploadSize + int64(binary.Size(encodedHistoryPart))
+		}
+
+		saveHistoryIteratorState(ctx, featureCatalog, historyIterator, part, &progress)
+	}
+
+	scope.AddCounter(metrics.HistoryArchiverTotalUploadSize, totalUploadSize)
+	scope.AddCounter(metrics.HistoryArchiverHistorySize, totalUploadSize)
+	scope.IncCounter(metrics.HistoryArchiverArchiveSuccessCount)
+	return
+}
+
+// Get is used to access an archived history. When context expires method should stop trying to fetch history.
+// The URI identifies the resource from which history should be accessed and it is up to the implementor to interpret this URI.
+// This method should thrift errors - see filestore as an example.
+func (h *historyArchiver) Get(ctx context.Context, URI archiver.URI, request *archiver.GetHistoryRequest) (*archiver.GetHistoryResponse, error) {
+
+	err := h.ValidateURI(URI)
+	if err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidURI.Error()}
+	}
+
+	if err := archiver.ValidateGetRequest(request); err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidGetHistoryRequest.Error()}
+	}
+
+	var token *getHistoryToken
+	if request.NextPageToken != nil {
+		token, err = deserializeGetHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, &shared.BadRequestError{Message: archiver.ErrNextPageTokenCorrupted.Error()}
+		}
+	} else if request.CloseFailoverVersion != nil {
+		_, historyhighestPart, historyCurrentPart, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *request.CloseFailoverVersion,
+			HighestPart:          *historyhighestPart,
+			NextBatchIdx:         *historyCurrentPart,
+		}
+	} else {
+		highestVersion, historyhighestPart, historyCurrentPart, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *highestVersion,
+			HighestPart:          *historyhighestPart,
+			NextBatchIdx:         *historyCurrentPart,
+		}
+	}
+
+	response := &archiver.GetHistoryResponse{}
+	response.HistoryBatches = []*shared.History{}
+	numOfEvents := 0
+	numOfBatches := 0
+
+outer:
+	for token.NextBatchIdx <= token.HighestPart {
```

Comment:
You will still need the `token.CurrentPart`. Current part is used to identify the current page number while next batch idx is used to trim history batches in the beginning of a page. So I think the code should look like the follows:

```
type getHistoryToken struct {
	CloseFailoverVersion int64
	CurrentPart          int
	HighestPart          int
	BatchIdxOffset       int
}

func Get() {
	// process next page token

	response := &archiver.GetHistoryResponse{}
	response.HistoryBatches = []*shared.History{}
	numOfEvents := 0

outer:
	for token.CurrentPart <= token.HighestPart {
		// download and decode file for the current part

		// trim the batches in the beginning based on token.BatchIdxOffset
		batches = batches[token.BatchIdxOffset:]

		for idx, batch := range batches {
			response.HistoryBatches = append(response.HistoryBatches, batch)
			token.BatchIdxOffset++
			numOfEvents += len(batch.Events)
			if numOfEvents >= request.PageSize {
				if idx == len(batches) - 1 {
					// handle the edge case where page size is meeted after adding the last batch
					token.BatchIdxOffset = 0
					token.CurrentPart++
				}
				break outer
			}
		}

		// reset the offset to 0 as we will read a new page
		token.BatchIdxOffset = 0
		token.CurrentPart++
	}
        // serializeToken and return
}
```

Please let me know your thoughts.

Reply:
Yes, I think that makes sense, especially if the pageSize is huge.